### DISCUSSION TITLE
File upload via GraphQL API

### DISCUSSION THREAD

**Discussion by sedirmohammed** [requarks/wiki#4456] [Help / Questions] [answered]
Hello!
Is it possible to upload files using the GraphQL API? If so, how would this work as an example? I haven't really found anything about this in the docs.

Kind regards

**Comment by xasz** [comment]
I have searched quite some time for this aswell.
You can create Folders / Rename Assets but could not figure out how to upload a file.

**Comment by NGPixel** [comment, marked as answer]
File uploads are not done via GraphQL, but a standard REST api endpoint at path `/u` using POST:

https://github.com/requarks/wiki/blob/main/server/controllers/upload.js#L13

**Reply by xasz** [reply]
Thanks for your response. That is a good point to look at. 
Tbh js is not my language that i very comfy to read.
I tried just posting a json body against it with the jwt api token, but the connection times out.

This is what i did in powershell. Not needing "help" there, just for calification:
```powershell
$fileBytes = ([IO.File]::ReadAllBytes($filePath))
$body = @{ ImportFile=$fileBytes } | ConvertTo-JSON
Invoke-RestMethod -Method POST -URI ("{0}/u" -f $uri) -ContentType "application/JSON" -Headers @{Authorization="Bearer $jwt"} -Body $body
```

Any chance you got an working curl or whatever command which works to understand how it works and how i do get the "wikijs" path to add it to a page.

What i try to accomplish is to upload a file to wikijs and then edit a page via graphql with the new link. (This part i got to work )

**Reply by xasz** [reply]
PS: I do got full acess on the wikijs host. If there is any chance i can just put the file in a upload folder or whatever would be no problem. I just need something i could call lateron via the wikijs page like: https://mywiki.wiki.com/givememyfile

**Reply by WERTBON1789** [reply]
So, I also wanted to upload files via this REST API, and I tinkered for a few hours

That's the basic Python script I came up with:
```
import requests

f = open('<path_to_file>', 'rb')

headers = {
    'Authorization': 'Bearer <token>'
}

url = 'http://wikijs.local/u'

files = (
    ('mediaUpload', (None, '{"folderId":1}')), # The number is the id of the asset folder, for the root of the page you need to give it a 0 I think, or you create a asset folder and specify it's id
    ('mediaUpload', ('filename.png', f, 'image/png'))
)

rq = requests.Request('POST', url, headers=headers, files=files) # print(rq.prepare().decode('utf-8')) to look at the request

result = requests.post(url, headers=headers, files=files) # Actually making the request

print(result.text)

f.close()
```

**Reply by WERTBON1789** [reply]
This creates a multipart/form-data request, which is in the same format as the Web interface sends it.
Found the request type and the layout of the data with my browser's Devtools

**Reply by xasz** [reply]
Hello @WERTBON1789 ,

thanks for shareing your code, i am struggeling to replicate it with powershell and i do not have any phyton runtime :)
Would it be possible to post a decoded request for me?

**Reply by crawc** [reply]
Is there a way to identify the newly uploaded file so it can be referenced or inserted into a wiki page?

I was able to upload the file but all I get is an "ok" text returned and I can see the file on the file system as a file hash, i.e. "cache/3f4d1fe1c04a1e58924da74e4604057f2dd62c59.dat"

**Update:** After looking at the GUI and uploading the files I see a GraphQL query (I'm still looking into this) I ran returning the complete list for the given folderId, then after the file is selected it seems to link or inline insert using a cleaned version of the filename. i.e. http://192.168.2.10:3000/some-file.pdf I uploaded the same file and it seems to replace the file and update the timestamp. so files with the same name seem to get overwritten without notice. 


Here is everything in Postman, that might help others too:
https://i.imgur.com/BlRUMvM.png

and the same as a CURL command:
`curl --location --request POST 'http://192.168.2.10:3000/u' \
--header 'Authorization: Bearer API_CODE_HERE' \
--form 'mediaUpload="{\"folderId\":3}"' \
--form 'mediaUpload=@"/C:/Users/jimmy/Nextcloud/HowTo_do_stuff.pdf"'`

**Reply by markus-96** [reply]
For those who are wondering how to accomplish this with aiohttp, here is my attempt. Sidenote: the user first logs in with `mutation{authentication{login(username, password, strategy)}}`, the cookie_jar is than stored with the retrieved cookie.

```python
import logging

import aiohttp
from aiohttp import ClientConnectorError, ServerTimeoutError, CookieJar
from yarl import URL


async def upload_asset(base_url: URL, cookie_jar: CookieJar, ssl_verification: bool, file_path: str, file_name: str, asset_folder_id: int, headers=None, verbose=False):
    if headers is None:
        headers = {}
    cookies = cookie_jar.filter_cookies(base_url)
    for cookie, value in cookies.items():
        if cookie == "jwt":
            headers.update({'Authorization': f"Bearer {value.value}"})

    async with aiohttp.ClientSession(cookie_jar=cookie_jar) as session:

        data = aiohttp.FormData(quote_fields=False)
        data.add_field('mediaUpload', f'{{"folderId":{asset_folder_id}}}')
        data.add_field('mediaUpload', open(file_path, 'rb'), filename=file_name,
                       content_type='image/jpeg')

        try:
            async with session.post(
                    url=base_url.with_path("u"),
                    ssl=ssl_verification,
                    data=data,
                    headers=headers
            ) as resp:
                print(f"status: {resp.status}")
                result = await resp.text()
                if verbose:
                    print(f"upload: {result}")
                return result
        except ClientConnectorError and ServerTimeoutError:
            if verbose:
                print("Exception in upload_asset")
            logging.exception("Exception in upload_asset")
            return {}

```
